src/direct/direct-export-buf.c

```diff
@@ -114,51 +115,68 @@ static bool import_to_cuda(NVDriver *drv, NVDriverImage *image, int bpc, int cha
 }
 
 BackingImage *direct_allocateBackingImage(NVDriver *drv, const NVSurface *surface) {
-    NVDriverImage driverImages[2] = { 0 };
-    int bpp = surface->format == cudaVideoSurfaceFormat_NV12 ? 8 : 16;
+    NVDriverImage driverImages[3] = { 0 };
+    const NVFormatInfo *fmtInfo;
+    const NVFormatPlane *p;
 
     BackingImage *backingImage = calloc(1, sizeof(BackingImage));
 
-    LOG("Allocating BackingImages: %p %dx%d", backingImage, surface->width, surface->height);
-    alloc_image(&drv->driverContext, surface->width, surface->height, 1, bpp, &driverImages[0]);
-    alloc_image(&drv->driverContext, surface->width>>1, surface->height>>1, 2, bpp, &driverImages[1]);
-
-    if (!import_to_cuda(drv, &driverImages[0], bpp, 1, &backingImage->cudaImages[0], &backingImage->arrays[0])) {
-        goto bail;
-    }
-    if (!import_to_cuda(drv, &driverImages[1], bpp, 2, &backingImage->cudaImages[1], &backingImage->arrays[1])) {
-        goto bail;
-    }
-
-    backingImage->fds[0] = driverImages[0].drmFd;
-    backingImage->fds[1] = driverImages[1].drmFd;
+    switch (surface->format)
+    {
+    case cudaVideoSurfaceFormat_P016:
+        if (surface->bitDepth == 10) {
+            backingImage->format = NV_FORMAT_P010;
+        } else {
+            backingImage->format = NV_FORMAT_P012;
+        }
+        break;
 
-    if (surface->format == cudaVideoSurfaceFormat_P016) {
+    case cudaVideoSurfaceFormat_YUV444_16Bit:
         if (surface->bitDepth == 10) {
-            backingImage->fourcc = DRM_FORMAT_P010;
+            backingImage->format = NV_FORMAT_Q410;
         } else {
-            backingImage->fourcc = DRM_FORMAT_P012;
+            backingImage->format = NV_FORMAT_Q412;
         }
-    } else {
-        backingImage->fourcc = DRM_FORMAT_NV12;
+        break;
+
+    case cudaVideoSurfaceFormat_YUV444:
+        backingImage->format = NV_FORMAT_444P;
+        break;
+    
+    default:
+        backingImage->format = NV_FORMAT_NV12;
+        break;
     }
 
-    backingImage->width = surface->width;
-    backingImage->height = surface->height;
+    fmtInfo = &formatsInfo[backingImage->format];
+    p = fmtInfo->plane;
+
+    LOG("Allocating BackingImages: %p %dx%d", backingImage, surface->width, surface->height);
+    for (int i = 0; i < fmtInfo->numPlanes; i++) {
```

Comment:
There's a bunch of places you're using an `int` in a for loop and comparing it to an unsigned value. Can you change those to `uint32_t`.

Reply:
I changed from `int` to `uint32_t` in loops with `numPlanes`.